src/api/app/models/workflow/step/submit_request.rb

```diff
@@ -0,0 +1,118 @@
+      submit_package
+    end
+    collect_artifacts
+  end
+
+  private
+
+  def collect_artifacts
+    Workflows::ArtifactsCollector.new(step: self, workflow_run_id: workflow_run.id, request_numbers_and_state_for_artifacts: @request_numbers_and_state_for_artifacts).call
+  end
+
+  def bs_request_description
+    step_instructions[:description] || workflow_run.event_source_message
+  end
+
+  def submit_package
+    # let possible running source services finish, before submitting the sources
+    Backend::Api::Sources::Package.wait_service(step_instructions[:source_project], step_instructions[:source_package])
+    bs_request_action = BsRequestAction.new(source_project: step_instructions[:source_project],
+                                            source_package: step_instructions[:source_package],
+                                            target_project: step_instructions[:target_project],
+                                            target_package: step_instructions[:target_package],
+                                            source_rev: source_package_revision,
+                                            type: 'submit')
+    bs_request = BsRequest.new(bs_request_actions: [bs_request_action],
+                               description: bs_request_description)
+    Pundit.authorize(@token.executor, bs_request, :create?)
+
+    begin
+      bs_request.save!
+    rescue MaintenanceHelper::MissingAction
+      raise 'Unable to submit, sources are unchanged'
+    rescue Project::Errors::UnknownObjectError
+      raise "Unable to submit: The source of package #{source_project_name}/#{source_package_name} is broken"
+    rescue APIError, ActiveRecord::RecordInvalid => e
+      raise e.message
+    rescue Backend::Error => e
+      raise e.summary
+    end
+
+    create_or_update_subscriptions(bs_request: bs_request)
+    (@request_numbers_and_state_for_artifacts["#{bs_request.state}"] ||= []) << bs_request.number
+    bs_request
+  end
+
+  def supersede_previous_and_submit_request
+    # Fetch current open submit request which are going to be superseded
+    # after the new sumbit request is created
+    requests_to_be_superseded = submit_requests_with_same_target_and_source
+    new_submit_request = submit_package
+
+    return if requests_to_be_superseded.blank?
+
+    requests_to_be_superseded.each do |submit_request|
+      # Authorization happens on model level
+      request = BsRequest.find_by_number!(submit_request.number)
+      request.change_state(newstate: 'superseded',
+                           reason: "Superseded by request #{new_submit_request.number}",
+                           superseded_by: new_submit_request.number)
+      (@request_numbers_and_state_for_artifacts["#{request.state}"] ||= []) << request.number
+    end
+  end
+
+  def revoke_submit_requests
+    return if submit_requests_with_same_target_and_source.blank?
```

Comment:
If there is nothing `submit_requests_with_same_target_and_source` will be `[]` and nothing will happen.

Reply:
removed